src/actions.py

```diff
@@ -43,26 +46,81 @@ def _block_keys_action(self, event: ops.ActionEvent) -> None:
         Args:
             event: the event triggering the original action.
         """
-        fingerprints = event.params["fingerprints"]
-        comment = event.params["comment"]
-        command = [
-            "/hockeypuck/bin/delete_keys.py",
-            "--fingerprints",
-            fingerprints,
-            "--comment",
-            comment,
-        ]
-        self._execute_action(event, command, leader_only=True)
-        command = ["/hockeypuck/bin/rebuild_prefix_tree.py"]
-        self._execute_action(event, command)
+        try:
+            if not self.charm.is_ready():
+                raise RuntimeError("Service not yet ready.")
+
+            input_fingerprints: str = event.params["fingerprints"]
+            comment: str = event.params["comment"]
+            fingerprints = input_fingerprints.split(",")
+
+            result = {}
+            fingerprints = [fingerprint.lower() for fingerprint in fingerprints]
+            for fingerprint in fingerprints:
+                if not re.fullmatch(r"[0-9A-Fa-f]{40}|[0-9A-Fa-f]{64}", fingerprint):
+                    result[fingerprint] = (
+                        "Invalid fingerprint format. "
+                        "Fingerprints must be 40 or 64 characters long and "
+                        "consist of hexadecimal characters only."
+                    )
+                    continue
+                response = requests.get(
+                    f"http://127.0.0.1:{HTTP_PORT}/pks/lookup?op=get&search=0x{fingerprint}",
+                    timeout=20,
+                )
+                if response.status_code not in (200, 404):
```

Comment:
I'd move this after the `if response.status_code == 404` and check on `not response.ok` to be on the safe side